File: 77.combinations.java
/*
 * @lc app=leetcode id=77 lang=java
 *
 * [77] Combinations
 */

// @lc code=start

import java.util.ArrayList;
import java.util.List;

class Solution {
    private List<List<Integer>> result;
    private List<Integer> path;
    private void backtracking(int n, int k, int startIndex) {
        if (path.size() == k) {
            result.add(new ArrayList<>(path));
            return;
        }
        for (int i = startIndex; i <= n - (k - path.size()) + 1; ++i) {
            path.add(i);
            backtracking(n, k, i + 1);
            path.removeLast();
        }
    }
    public List<List<Integer>> combine(int n, int k) {
        result = new ArrayList<>();
        path = new ArrayList<>();
        backtracking(n, k, 1);
        return result;
    }
}
// @lc code=end

File: 77.combinations.cpp
/*
 * @lc app=leetcode id=77 lang=cpp
 *
 * [77] Combinations
 */

// @lc code=start

#include <vector>

using namespace std;

class Solution {
public:

    vector<vector<int>> results;
    vector<int> path;

    vector<vector<int>> combine(int n, int k) {
        backtracking(n, k, 1);
        return results;
    }

    void backtracking(int n, int k, int startIndex) {
        if (path.size() == k) {
            results.push_back(path);
        }
        for (int ix = startIndex; ix <= n; ++ix) {
            if (ix + (k - path.size()) - 1 > n) break;
            path.push_back(ix);
            backtracking(n, k, ix + 1);
            path.pop_back();
        }
    }
};
// @lc code=end
